### DISCUSSION TITLE
【组件开发】是否有监听播放器模式切换的API

### DISCUSSION THREAD

**Discussion by ZiuChen** [the1812/Bilibili-Evolved#4178] [提问 / Q&A] [answered]
目前，启用[播放器置顶](https://github.com/the1812/Bilibili-Evolved/blob/preview-features/registry/lib/components/style/player-on-top-new/index.ts)这个组件会带来副作用：UP主卡片被挪到了弹幕列表的下面
> <img width="50%" alt="image" src="https://user-images.githubusercontent.com/64892985/235945420-636b87f8-37c5-430e-99d4-a31f176c3c4f.png">
这是因为：如果不使用`moveElementAfter`移动元素，在宽屏模式下UP主卡片会被弹幕列表的`margin-top`顶到上面
> <img width="50%" alt="image" src="https://user-images.githubusercontent.com/64892985/235945818-5251c620-d9bc-434d-b972-a895a1b42f83.png">

而我希望能够在启用播放器置顶的功能的同时，保持之前UP主卡片在上、弹幕列表在下的布局
对播放器模式切换做监听，在模式切换后，通过改变它们的CSS属性来保证布局不会出错

```ts
// registry/lib/components/style/player-on-top-new/index.ts
  const author = document.querySelector('#v_upinfo').parentElement
  const danmuku = document.querySelector('#danmukuBox') as HTMLDivElement

  // moveElementAfter(author, danmuku)
  onVideoModeChange((mode) => {
    // 宽屏模式 重写元素的CSS属性
    if(mode === 'wide-mode') {
      author.style.marginTop = '631px'
      danmuku.style.marginTop = ''
    }
  })
```

我想知道是否有像`onVideoModeChange`这样的API可以实现类似的操作，或者是否有性能更好的方式来对用户操作做监听？

**Comment by the1812** [comment, marked as answer]
一般是根据 body 上的 class 判断的, 比如宽屏模式下会带有 `player-mode-widescreen` 的 class, 正好你需要的也是修改 CSS

**Reply by ZiuChen** [reply]
谢谢！刚刚简单测试了一下，`body`上类名的响应式似乎还有点问题：

- 第一次普通切换到宽屏 会响应
- 宽屏切到网页全屏 会响应
- 后续普通与宽屏的切换 **不会响应**

简单提了个[PR](https://github.com/the1812/Bilibili-Evolved/pull/4188)，劳驾看看是否合适，合适的话我将在此基础上继续改进播放器置顶组件

[bpx.ts 源码](https://github.com/the1812/Bilibili-Evolved/blob/d8dc95f852978b90ed9e0dfc726f55e05c7baa7d/src/components/video/player-adaptor/bpx.ts#LL12C34-L12C34)

```ts
// bpx.ts
...
  attributes(bpxContainer, () => {
    const dataScreen = bpxContainer.getAttribute('data-screen')
    const prefix = 'player-mode-'
    const enumList = ['normal', 'wide', 'web', 'full'].map(it => `${prefix}${it}`)

    // clear all class
    document.body.classList.remove(...enumList)

    // add class
    document.body.classList.add(dataScreen !== 'normal' ? `${prefix}${dataScreen}` : '')
  })
...
```

**Reply by the1812** [reply]
应该没问题的, 现在都是 bpx 播放器了, 以前这样写是为了兼容老版本的播放器
